File: framework/lib/command.dart
part of uxml;

/**
 * Manages a command and associated event routing. Implements command
 * pattern (http://en.wikipedia.org/wiki/Command_pattern) for framework.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class Command extends UxmlElement {

  static ElementDef commandElementDef;
  /** Command event definition */
  static EventDef commandEvent = null;

  /** Enabled property */
  static PropertyDefinition enabledProperty = null;

  Command(String commandId) {
    id = commandId;
  }

  /**
   * Walks up view tree and sends command event to each view node and
   * finally executes listeners on command instance.
   *
   * sourceEvent contains the initial event such input events that triggered
   * the command.
   */
  void execute(EventArgs sourceEvent, [Object params = null]) {
    CommandEventArgs cmdArgs = new CommandEventArgs(this,
        sourceEvent.source, params);
    cmdArgs.event = commandEvent;
    if (sourceEvent.source is UIElement) {
      UIElement viewNode = sourceEvent.source;
      viewNode.routeEvent(cmdArgs);
    }
    notifyListeners(commandEvent, cmdArgs);
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => commandElementDef;

  /** Registers component. */
  static void registerCommand() {
    commandEvent = new EventDef("Command", Route.BUBBLE);
    enabledProperty = ElementRegistry.registerProperty("enabled",
        PropertyType.BOOL, PropertyFlags.NONE, null, true);
    commandElementDef = ElementRegistry.register("Command",
      null, [enabledProperty], [commandEvent]);
  }
}

/**
 * Defines command event arguments.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class CommandEventArgs extends EventArgs {

  Command _cmd;
  /** Sets or returns command parameters. */
  Object parameters = null;

  CommandEventArgs(Command command, Object source, Object params) :
      super(source) {
    _cmd = command;
    parameters = params;
  }

  /** Returns command. */
  Command get command => _cmd;
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Command.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventDefinition;
import com.hello.uxml.tools.framework.events.EventManager;

import java.util.EnumSet;

/**
 * Manages a command and associated event routing. Implements command
 * pattern (http://en.wikipedia.org/wiki/Command_pattern) for framework.
 *
 * @author ferhat
 */
public class Command extends UxmlElement {

  private String id;

  /** Click event definition */
  public static EventDefinition commandEvent = EventManager.register(
      "Command", UIElement.class, CommandEventArgs.class, UIElement.class, null);

  /** Enabled property definition */
  public static PropertyDefinition enabledPropDef = PropertySystem.register("Enabled",
      Boolean.class, UIElement.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  public Command() {
    id = "";
  }

  public Command(String commandId) {
    id = commandId;
  }

  /** Returns id of command. */
  @Override
  public String getId() {
    return id;
  }

  /** Sets or returns whether command is enabled. */
  public void setEnabled(boolean val) {
    setProperty(enabledPropDef, val);
  }

  public boolean getEnabled() {
    return ((Boolean) getProperty(enabledPropDef)).booleanValue();
  }

  public void execute(EventArgs e, Object param) {
    CommandEventArgs data = new CommandEventArgs(this, e.getSource(), param);
    data.setEvent(commandEvent);
    if (e.getSource() instanceof UxmlElement) {
      ((UxmlElement) e.getSource()).notifyListeners(commandEvent, data);
    }
    notifyListeners(commandEvent, data);
  }
}
